File: backend/app/schemas/role.py
"""
Role schemas for request/response.
"""
from typing import Optional, List
from pydantic import BaseModel, Field


class RoleBase(BaseModel):
    """Base role schema."""
    name: str = Field(..., description="角色名称")
    code: str = Field(..., description="角色编码")
    sort: int = Field(0, description="排序")
    status: int = Field(1, description="状态:0禁用,1正常")
    data_scope: int = Field(1, description="数据权限:1全部,2本部门及以下,3本部门,4仅本人")


class RoleCreate(RoleBase):
    """Schema for creating role."""
    permission_ids: List[int] = Field(default_factory=list, description="权限ID列表")


class RoleUpdate(BaseModel):
    """Schema for updating role."""
    name: Optional[str] = None
    code: Optional[str] = None
    sort: Optional[int] = None
    status: Optional[int] = None
    data_scope: Optional[int] = None
    permission_ids: Optional[List[int]] = None


class RoleResponse(RoleBase):
    """Schema for role response."""
    id: int
    tenant_id: int
    
    class Config:
        from_attributes = True


class RoleWithPermissions(RoleResponse):
    """Role with permission list."""
    permissions: List["PermissionResponse"] = Field(default_factory=list, description="权限列表")
    
    class Config:
        from_attributes = True


# Forward reference for PermissionResponse
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from app.schemas.permission import PermissionResponse

File: frontend/src/types/role.ts
/**
 * 角色相关类型定义
 */
import { PaginationParams } from './api';

/**
 * 角色基础类型
 */
export interface Role {
  id: string;
  name: string;
  code: string;
  sort: number;
  status: number;
  data_scope: number;
  tenant_id: string;
  created_at: string;
  updated_at?: string;
}

/**
 * 创建角色类型
 */
export interface RoleCreate {
  name: string;
  code: string;
  sort?: number;
  status?: number;
  data_scope?: number;
  permission_ids?: string[];
  department_ids?: string[]; // 当 data_scope=5 时使用
}

/**
 * 更新角色类型
 */
export interface RoleUpdate {
  name?: string;
  code?: string;
  sort?: number;
  status?: number;
  data_scope?: number;
  permission_ids?: string[];
  department_ids?: string[]; // 当 data_scope=5 时使用
}

/**
 * 角色列表查询参数
 */
export interface RoleListParams extends PaginationParams {
  name?: string;
  code?: string;
  status?: number;
}

/**
 * 权限树节点类型
 */
export interface PermissionTreeNode {
  id: string;
  name: string;
  code: string;
  type: number; // 1目录,2菜单,3按钮
  parent_id?: string;
  sort: number;
  status: number;
  tenant_id: string;
  children?: PermissionTreeNode[];
  checked?: boolean; // 前端使用，标记是否选中
}

